web/libs/editor/src/lib/AudioUltra/Analysis/MelBanks.ts

```diff
@@ -0,0 +1,163 @@
+    }
+  }
+
+  /**
+   * Applies the pre-calculated Mel filter bank to a linear power spectrum.
+   *
+   * @param linearSpectrum - The input linear power spectrum (magnitudes).
+   * @returns The spectrum converted to the Mel scale.
+   */
+  applyFilterbank(linearSpectrum: Float32Array): Float32Array {
+    if (this.filterbank.length === 0 || this.numBands === 0) {
+      console.warn("Mel filter bank not initialized or invalid.");
+      // Return an empty or zero-filled array matching the expected band count
+      return new Float32Array(this.numBands).fill(0);
+    }
+
+    // Ensure the filter bank bin count matches linear spectrum length
+    if (this.filterbank.length > 0 && this.filterbank[0].length !== linearSpectrum.length) {
+      console.error(
+        `Mel filter bank bin count (${this.filterbank[0].length}) does not match linear spectrum length (${linearSpectrum.length}). Recreate MelBanks instance.`,
+      );
+      return new Float32Array(this.numBands).fill(0);
+    }
+
+    const melSpectrum = new Float32Array(this.numBands).fill(0);
+    for (let i = 0; i < this.numBands; i++) {
+      const filter = this.filterbank[i];
+      // filter bank[0].length should equal linearSpectrum.length due to check above
+      for (let j = 0; j < filter.length; j++) {
+        melSpectrum[i] += linearSpectrum[j] * filter[j];
+      }
+      // Add a small epsilon to avoid log(0) if further log scaling is applied later
+      melSpectrum[i] = melSpectrum[i] > 0 ? melSpectrum[i] : 1e-10;
+    }
+
+    return melSpectrum;
+  }
+
+  /**
+   * Converts frequency in Hz to Mel scale.
+   */
+  private hzToMel(hz: number): number {
+    return 2595 * Math.log10(1 + hz / 700);
+  }
+
+  /**
+   * Converts Mel scale value back to frequency in Hz.
+   */
+  private melToHz(mel: number): number {
+    return 700 * (Math.pow(10, mel / 2595) - 1);
+  }
+
+  /**
+   * Creates a Mel filter bank matrix.
+   */
+  private createMelFilterbank(sampleRate: number, linearBinCount: number, numBands: number): number[][] {
+    // Basic parameter validation
+    if (numBands <= 0 || linearBinCount <= 1 || sampleRate <= 0) {
+      console.warn("Invalid parameters for Mel filterbank creation.");
+      return [];
+    }
+
+    const lowFreqMel = 0;
+    const highFreqMel = this.hzToMel(sampleRate / 2);
+
+    // Check for degenerate Mel range
+    if (lowFreqMel >= highFreqMel) {
+      console.warn("Min Mel frequency is not less than Max Mel frequency.");
+      return [];
+    }
+
+    const melPoints = new Float32Array(numBands + 2);
+    const melStep = (highFreqMel - lowFreqMel) / (numBands + 1);
+
+    // Check for a non-positive Mel step
+    if (melStep <= 0) {
+      console.warn("Calculated Mel step is not positive.");
+      return [];
+    }
+
+    // Create evenly spaced points in a Mel scale
+    for (let i = 0; i < numBands + 2; i++) {
+      melPoints[i] = lowFreqMel + i * melStep;
```

Comment:
Given how this was being used, we can just retain the TypedArray by removing the intermediary step and secondary map call.
```suggestion
      hzPoints[i] = this.melToHz(lowFreqMel + i * melStep);
```